From 7e32832b684e2545a86b02635a93f2fe7b639759 Mon Sep 17 00:00:00 2001
From: Ian Douglas Scott <idscott@system76.com>
Date: Wed, 30 Dec 2020 08:38:44 -0800
Subject: [PATCH] mouse: Add "Disable While Typing" toggle for touchpad

https://gitlab.gnome.org/GNOME/gnome-control-center/-/issues/125

As mentioned there, this is exposed in Gnome Tweaks. It is on by
default, but completely break touchpad usage of certain kinds of games
and other software requiring simultaneous use of the mouse and keyboard.

I like the idea in the issue of automatically detecting games, but I'm
not sure how/if it can be done reliably (and it's not an issue only for
games). In lieu of that, it seems good to expose the setting here.
---
 panels/mouse/cc-mouse-panel.c  |  6 ++++
 panels/mouse/cc-mouse-panel.ui | 52 ++++++++++++++++++++++++++++++++++
 2 files changed, 58 insertions(+)

--- a/panels/mouse/cc-mouse-panel.c
+++ b/panels/mouse/cc-mouse-panel.c
@@ -36,6 +36,7 @@
 {
   CcPanel            parent_instance;
 
+  GtkSwitch         *disable_while_typing_switch;
   GtkListBoxRow     *edge_scrolling_row;
   GtkSwitch         *edge_scrolling_switch;
   GtkListBox        *general_listbox;
@@ -329,6 +330,10 @@
                    self->edge_scrolling_switch, "state",
                    G_SETTINGS_BIND_GET);
 
+  g_settings_bind (self->touchpad_settings, "disable-while-typing",
+                   self->disable_while_typing_switch, "active",
+                   G_SETTINGS_BIND_DEFAULT);
+
   setup_touchpad_options (self);
 }
 
@@ -449,6 +454,7 @@
 
   gtk_widget_class_set_template_from_resource (widget_class, "/org/gnome/control-center/mouse/cc-mouse-panel.ui");
 
+  gtk_widget_class_bind_template_child (widget_class, CcMousePanel, disable_while_typing_switch);
   gtk_widget_class_bind_template_child (widget_class, CcMousePanel, edge_scrolling_row);
   gtk_widget_class_bind_template_child (widget_class, CcMousePanel, edge_scrolling_switch);
   gtk_widget_class_bind_template_child (widget_class, CcMousePanel, general_listbox);
--- a/panels/mouse/cc-mouse-panel.ui
+++ b/panels/mouse/cc-mouse-panel.ui
@@ -366,6 +366,28 @@
                             </child>
                           </object>
                         </child>
+                        <child>
+                          <object class="HdyActionRow">
+                            <property name="visible">True</property>
+                            <property name="can_focus">True</property>
+                            <property name="activatable">false</property>
+                            <property name="title" translatable="yes">Disable while typing</property>
+                            <child>
+                              <object class="GtkSwitch" id="disable_while_typing_switch">
+                                <property name="visible">True</property>
+                                <property name="can_focus">True</property>
+                                <property name="halign">end</property>
+                                <property name="valign">center</property>
+                              </object>
+                              <packing>
+                                <property name="left_attach">1</property>
+                                <property name="top_attach">0</property>
+                                <property name="width">1</property>
+                                <property name="height">2</property>
+                              </packing>
+                            </child>
+                          </object>
+                        </child>
                       </object>
                     </child>
                   </object>
